x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Repeat.java

```diff
@@ -0,0 +1,107 @@
+import org.elasticsearch.compute.ann.Evaluator;
+import org.elasticsearch.compute.operator.EvalOperator.ExpressionEvaluator;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.function.OptionalArgument;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.core.type.DataTypes;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+
+public class Repeat extends EsqlScalarFunction implements OptionalArgument {
+
+    private final Expression str;
+    private final Expression number;
+
+    @FunctionInfo(
+        returnType = "keyword",
+        description = "Returns a string constructed by concatenating `string` with itself the specified `number` of times.",
+        examples = @Example(file = "string", tag = "repeat")
+    )
+    public Repeat(
+        Source source,
+        @Param(name = "string", type = { "keyword", "text" }, description = "String expression.") Expression str,
+        @Param(name = "number", type = { "integer" }, description = "Number times to repeat.") Expression number
+    ) {
+        super(source, Arrays.asList(str, number));
+        this.str = str;
+        this.number = number;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isString(str, sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return isType(number, dt -> dt == DataTypes.INTEGER, sourceText(), SECOND, "integer");
+    }
+
+    @Override
+    public boolean foldable() {
+        return str.foldable() && number.foldable();
+    }
+
+    @Evaluator
+    static BytesRef process(BytesRef str, int number) {
+        if (str == null) {
+            return null;
+        }
+        if (number < 0) {
+            throw new IllegalArgumentException("Number parameter cannot be negative, found [" + number + "]");
```

Comment:
Sounds good, I changed it to return null on negative `number` args and add the warnExceptions. For the case where the 1MB upper bound field size is exceeded, it throws a ` EsqlClientException` rather than adding to warnExceptions since we probably don't want to fail quietly in this situation. What do you think?

Reply:
I think in both cases we want to fail with the warning. At least, that's the recommendation that we've been working under.

I do think you can move both of those tests into the non-fixed process function and check them up front. Up front it's quite ok to throw an IllegalArgumentException an fail the query.